### ISSUE TITLE
Prevent the success notification appearing when returning from OAuth in the GA4 Activation Banner flow

### ISSUE THREAD

**Issue by techanvil** [google/site-kit-wp#5837]
## Feature Description

As described in https://github.com/google/site-kit-wp/issues/5282#issuecomment-1246528980, at present, when returning from OAuth to the GA4 Activation Banner, the Success Notification is displayed, interrupting the desired GA4 Activation Banner flow.

We need to prevent the Success Notification from being displayed in this particular case.

Here is a screenshot of the Success Notification, shown when the URL contains the query param `notification=authentication_success`.
![image.png](https://images.zenhubusercontent.com/6133937d41f397090b8344a7/93aed54c-3b09-474b-b634-e733cd687275)

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* When returning from OAuth to the GA4 Activation Banner, the Success Notification should not be displayed, and the GA4 Activation Banner should be shown in its expected state.

## Implementation Brief

We will add a `skipNotification` query param to the `redirectURL` used to return from OAuth to the GA4 Activation Banner. When `skipNotification` is present in the `redirectURL`, we will _not_ add the `notification=authentication_success` query param to said redirect URL.

- Finish off the PR

### Test Coverage

- Fix any failing tests.
- Add some PHPUnit test coverage for the additions to `OAuth_Client`.

## QA Brief

* 

## Changelog entry

*

**Comment by aaemnnosttv** [comment]
Just a thought: I don't think an added query parameter is necessary if we render the GA4 banner above the congrats banner as only one will be shown.

**Comment by techanvil** [comment]
> Just a thought: I don't think an added query parameter is necessary if we render the GA4 banner above the congrats banner as only one will be shown.

That's an interesting point. However, the congrats banner will still eventually show up once we dismiss the GA4 banner, I would imagine it would be preferable to not show it at all as the GA4 banner has its own success message...

**Comment by tofumatt** [comment]
Yeah, I was about to say the same thing as Evan: it'd be better to reorder the notification banners here and that would save us needing to add the query param, which seems slightly awkward to me.

The ACs here look good, but I think the IB approach is too complicated. I think it's okay to hide the GA4 banner and then still show the Site Kit one… or maybe even… when rendering the GA4 Banner, have it automatically hide the Congrats banner in a `useMount`. 🤔

**Comment by aaemnnosttv** [comment]
> * The `prompt=consent` argument is added by OAuth consent screen when it redirects the user back to WordPress, so we can use it as an indicator that the user has just finished providing additional scopes and we don't need to display the congrats notification.

@eugene-manuilov this looks like it would work but I'm not sure we should change it in all cases because we _are_ coming back with successful authentication 🤔

Two thoughts I have as possible solutions to consider:

- Push/replace a different `notification` parameter onto the URL when submitting changes - this would then be passed to the [`redirectURL` provided in the permissions modal](https://github.com/google/site-kit-wp/blob/540926060f5a0caec43c1925858d313ea94e561f/assets/js/components/PermissionsModal/AuthenticatedPermissionsModal.js#L44). That way `SetupSuccessBannerNotification` wouldn't be rendered at all and also, we could use this with a GA4-specific value to make sure the setup banner is forced to display when returning.
- Set a value in the UI store to override the `notification` in the query parameter or simply not render the `SetupSuccessBannerNotification`. This store will be snapshotted before going to oAuth and restored on return, so we could use a simple value in state as a single-use override.

Thoughts?

**Comment by techanvil** [comment]
> * Push/replace a different `notification` parameter onto the URL when submitting changes - this would then be passed to the [`redirectURL` provided in the permissions modal](https://github.com/google/site-kit-wp/blob/540926060f5a0caec43c1925858d313ea94e561f/assets/js/components/PermissionsModal/AuthenticatedPermissionsModal.js#L44). That way `SetupSuccessBannerNotification` wouldn't be rendered at all and also, we could use this with a GA4-specific value to make sure the setup banner is forced to display when returning.

This would still need a bit of logic on the PHP side, as the `notification` query param is currently unconditionally added to the redirect URL in PHP territory, using `add_query_arg` which overrides an existing param. So we'd need to add a condition that checks for the existence of `notification` and only adds if it doesn't exist. This would need doing in the same places I modified the PHP in my initial proposed [fix](https://github.com/google/site-kit-wp/pull/5838/files).

Also, the GA4 banner is already set up to display again on return, in the same state it was left in, so we don't need to worry about that aspect.

> * Set a value in the UI store to override the `notification` in the query parameter or simply not render the `SetupSuccessBannerNotification`. This store will be snapshotted before going to oAuth and restored on return, so we could use a simple value in state as a single-use override.

This could work... Although, I would prefer _not_ to have to store something in local state, which could potentially go out of sync (say the user closes the window while on the OAuth screen, and then comes back to Site Kit later and enters OAuth for a different reason, the "Congrats" notification could be suppressed when it shouldn't be).

**Comment by eugene-manuilov** [comment]
Thanks, @techanvil and @aaemnnosttv. I have update IB. Let me know if it looks good to you.

**Comment by techanvil** [comment]
Thanks @eugene-manuilov. This looks good, just a couple of small points:

- The `ga4_setup` value is essentially a placeholder here for "don't show a notification", maybe we should use a value like `none` or `do_not_show` instead? To my eyes, `notification=ga4_setup` makes it look like there _should_ be a notification displayed...
- From my previous testing - we will need the same change in `OAuth_Client::get_authentication_url` as in `OAuth_Client::authorize_user`, to avoid adding the `notification` query param if it's already present on the redirect URL.

**Comment by techanvil** [comment]
Thanks @eugene-manuilov. I'm happy enough to stick with `ga4_setup` in that case - we can always revisit it later.

**Comment by techanvil** [comment]
IB :white_check_mark:

**Comment by wpdarren** [comment]
### QA Update: ✅ 

**Verified:**

- The "Congrats..." banner should no longer appear when returning from OAuth to the GA4 Activation Banner.
- The GA4 Activation Banner will no longer appear in view-only mode. 
- Dashboard Sharing was tested with an admin and editor user role to make sure that the banner did not appear. 

![image](https://user-images.githubusercontent.com/73545194/190839771-b35eb2e7-cb15-4235-b69a-18fe5872996e.png)

https://user-images.githubusercontent.com/73545194/190839784-ba947d91-a861-40ed-abb2-fc5a3f82e303.mp4

**Comment by wpdarren** [comment]
### QA Update: ❌ 

This does pass the QAB, but there's an issue and I am unsure if the cause is this ticket or #5282.

The GA4 flow (with and without oAuth) goes through as expected. You see the `You Successfully Set up a GA4 property`  notification, and when checking Google Analytics the new GA4 property is created. BUT when you go to Site Kit Settings, the `Complete Google Analytics 4` button appears as if it isn’t set up.  

I am not entirely sure which ticket has caused this issue so have sent both back to execution for investigation.

**Comment by hussain-t** [comment]
@wpdarren. I have fixed the above issue in this follow-up [PR](https://github.com/google/site-kit-wp/pull/5866) for #5282.

**Comment by techanvil** [comment]
As discussed on Slack the screen shifting issue is no longer occurring so moving this back to QA.

**Comment by wpdarren** [comment]
### QA Update: ❌ 

@hussain-t apologies, I possibly should have picked this up before - not 100% sure if I tested this or not, but on the entity Site Kit dashboard screen, the GA4 banner is appearing.  It's not on the main dashboard. 

![image](https://user-images.githubusercontent.com/73545194/191235842-f979c0f8-56e9-4c93-a5aa-c57174d513a4.png)

**Comment by wpdarren** [comment]
### QA Update: ✅ 

**Verified:**

- The "Congrats..." banner no longer appears when returning from OAuth to the GA4 Activation Banner.
- The GA4 Activation Banner no longer appears in view-only mode.
- Dashboard Sharing was tested with an admin and editor user role to make sure that the banner did not appear.

Note: on a second test, I did notice that the banner does appear on the entity dashboard. I will create a ticket for this and class it as a known issue for the bug bash.
